packages/squiggle-lang/src/rescript/Reducer/Reducer_Expression/Reducer_Expression.res

```diff
@@ -38,50 +39,137 @@ let parse_ = (expr: string, parser, converter): result<t, errorValue> =>
+            let rNewExpression = replaceSymbols(expression, bindings)
+            rNewExpression->Result.map(newExpression =>
+              Belt.Map.String.set(bindings, aSymbol, newExpression)->T.EBindings
+            )
+          }
+        | _ => REAssignmentExpected->Error
+        }
+      }
+
+      let doBindExpression = (expression: t, bindings: T.bindings) => {
+        switch expression {
+        | T.EList(list{T.EValue(EvCall("$let")), ..._}) => REExpressionExpected->Error
+        | _ => replaceSymbols(expression, bindings)
+        }
+      }
+
+      switch list {
+      | list{T.EValue(EvCall("$$bindings"))} => bindings->EBindings->Ok
+
+      | list{T.EValue(EvCall("$$bindStatement")), T.EBindings(bindings), statement} =>
+        doBindStatement(statement, bindings)
+      | list{T.EValue(EvCall("$$bindExpression")), T.EBindings(bindings), expression} =>
+        doBindExpression(expression, bindings)
+      | _ => list->T.EList->Ok
+      }
+    }
+
+    list->dispatchMacroCall(bindings)
   }
 
-/*
-  Recursively evaluate/reduce the code tree
-*/
-let rec reduceExpression = (expression: t, bindings): result<expressionValue, 'e> =>
-  switch expression {
-  | T.EValue(value) => value->Ok
-  | T.EList(list) => {
-      let racc: result<list<expressionValue>, 'e> = list->Belt.List.reduceReverse(Ok(list{}), (
-        racc,
-        each: expression,
-      ) =>
-        racc->Result.flatMap(acc => {
-          each
-          ->reduceExpression(bindings)
-          ->Result.flatMap(newNode => {
-            acc->Belt.List.add(newNode)->Ok
+  let rec seekMacros = (expression: t, bindings: T.bindings): result<t, 'e> =>
+    switch expression {
+    | T.EValue(value) => expression->Ok
+    | T.EList(list) => {
+        let racc: result<list<t>, 'e> = list->Belt.List.reduceReverse(Ok(list{}), (
+          racc,
+          each: expression,
+        ) =>
+          racc->Result.flatMap(acc => {
+            each
+            ->seekMacros(bindings)
+            ->Result.flatMap(newNode => {
+              acc->Belt.List.add(newNode)->Ok
+            })
           })
-        })
-      )
-      racc->Result.flatMap(acc => acc->reduceValueList)
+        )
+        racc->Result.flatMap(acc => acc->doMacroCall(bindings))
+      }
     }
-  }
+
+  let rec reduceExpandedExpression = (expression: t): result<expressionValue, 'e> =>
+    switch expression {
+    | T.EValue(value) => value->Ok
+    | T.EList(list) => {
+        let racc: result<list<expressionValue>, 'e> = list->Belt.List.reduceReverse(Ok(list{}), (
+          racc,
+          each: expression,
+        ) =>
+          racc->Result.flatMap(acc => {
+            each
+            ->reduceExpandedExpression
+            ->Result.flatMap(newNode => {
+              acc->Belt.List.add(newNode)->Ok
+            })
+          })
+        )
+        racc->Result.flatMap(acc => acc->reduceValueList)
+      }
```

Comment:
It seems kind of janky for this to only handle 2 of the several possible expresssions. Similar with ``seekMacros``. 

I guess this would be a good case for polymorphic variants or similar. (As in, these functions only accept polymorphic variants for EValue / EList, and callers convert to those variants)

Reply:
You are right. I missed that because somehow it passed all the test cases. I guess that also means that the other cases are never ever arriving there.

Reply:
But I object to polymorphic variants. They are true danger.

Reply:
umuro/squiggle#53

Reply:
Also todo: Checking warning settings. Regardless of test results I should have seen a flashing warning! I did not!